### PR TITLE
TYP: mostly io, plotting

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

cc @simonjayhawkins can i get your thoughts on two remaining mypy complaints

```
pandas/io/excel/_base.py:795: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable)  [attr-defined]
pandas/io/stata.py:892: error: unused 'type: ignore' comment
```

The excel one is because it doesn't recognize 

```
@property
@abstractmethod
def supported_extensions(self) -> Tuple[str, ...]:
```

as returning `Tuple[str, ...]`.  I expect if that were resolved it would still complain bc subclasses just pin the tuple as a class attribute instead of a property (easy to make a property, but much more verbose which im not wild about)

The stata unused type:ignore becomes used once we remove stata from the setup.cfg exclusion.  The only working way I've found to make mypy accept `TYPE_MAP: List[Union[int, str]] = list(range(251)) + list("abcd")` is with

```
TYPE_MAP: List[Union[int, str]] = cast(List[Union[int, str]], list(range(251))) + list("abcd")
```

which i find much less clear than the `# type:ignore`d version.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
punted on issues in the OP, so this should be ready

**Comment by simonjayhawkins** [comment]
> cc @simonjayhawkins can i get your thoughts on two remaining mypy complaints
> 
> ```
> pandas/io/excel/_base.py:795: error: "Callable[[ExcelWriter], Tuple[str, ...]]" has no attribute "__iter__" (not iterable)  [attr-defined]
> ```

see https://github.com/pandas-dev/pandas/pull/28339#discussion_r339136692

**Comment by jbrockmendel** [comment]
Makes sense, thanks.  Are you OK with the solution here?

**Comment by simonjayhawkins** [comment]
> Makes sense, thanks. Are you OK with the solution here?

from the previous discussion, https://github.com/pandas-dev/pandas/pull/28339#discussion_r339180390

> Sure you could just remove the abstract requirement altogether for supported_extensions and make it a class variable, but that would lose the abstract nature of it.

maybe `type: ignore` for now and reference the mypy issue, https://github.com/python/mypy/issues/7760

The crux of the issue seems to be https://github.com/python/mypy/issues/7760#issuecomment-544473268

> I guess my real issue then is Python language's lack of notation for abstract class attributes.

The code does work. so we shouldn't really need to change it.

File: pandas/plotting/_matplotlib/tools.py [hunk 1]
```diff
@@ -401,11 +401,11 @@ def handle_shared_axes(
                     _remove_labels_from_axis(ax.yaxis)
 
 
-def flatten_axes(axes: Union["Axes", Sequence["Axes"]]) -> Sequence["Axes"]:
+def flatten_axes(axes: Union["Axes", Sequence["Axes"]]) -> np.ndarray:
     if not is_list_like(axes):
         return np.array([axes])
     elif isinstance(axes, (np.ndarray, ABCIndexClass)):
-        return axes.ravel()
+        return np.asarray(axes).ravel()
```
Comment (simonjayhawkins) [inline pandas/plotting/_matplotlib/tools.py:408]: The issue here is that the types are not refined by `ABCIndexClass`?

for a numpy array, this impacts perf.
```
%timeit np.asarray(axes).ravel()
# 724 ns ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit axes.ravel()
# 303 ns ± 9.35 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Is this important?

maybe split the elif into two or ignore till we resolve the ABC issue.

Comment (simonjayhawkins) [reply pandas/plotting/_matplotlib/tools.py:408]: or can we cast to Union[np.ndarray, "Index"]

Comment (jbrockmendel) [reply pandas/plotting/_matplotlib/tools.py:408]: Index.ravel behavior was just deprecated, so the np.asarray retains the current behavior

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel lgtm

**Pull request merged**